site/search/searchUtils.tsx

```diff
@@ -350,37 +397,231 @@ export async function querySearch(
         (country) => country.name
     )
-    const lastWord = query.split(" ").at(-1) ?? ""
-    const countries = FuzzySearch.withKey(
+
+    const queryWords = query.trim().split(/\s+/)
+
+    if (!queryWords.length || queryWords[0] === "") {
+        return {
+            suggestions: [],
+            unmatchedQuery: "",
+        }
+    }
+
+    const searchResults = findMatches(
+        queryWords,
         allCountryNames,
-        (country) => country,
+        allTopics,
+        selectedCountryNames,
+        selectedTopics,
         sortOptions
     )
-        .search(lastWord)
-        .filter((country) => !appliedFilters.selectedCountryNames.has(country))
-        .map((name) => ({ name, type: "country" as const }))
-    const tags = FuzzySearch.withKey(allTopics, (topic) => topic, sortOptions)
-        .search(lastWord)
-        .slice(0, 3)
-        .filter((topic) => !appliedFilters.selectedTopics.has(topic))
-        .map((name) => ({ name, type: "topic" as const }))
-
-    return [...countries, ...tags]
+
+    const unmatchedQuery = queryWords
+        .slice(0, searchResults.matchStartIndex)
+        .join(" ")
+
+    const countryMatches = searchResults.countryResults.map((result) => ({
+        filter: createCountryFilter(result.name),
+        score: result.score,
+    }))
+
+    const topicMatches = searchResults.topicResults.map((result) => ({
+        filter: createTopicFilter(result.name),
+        score: result.score,
+    }))
+
+    const allMatches = [...countryMatches, ...topicMatches]
+
+    const [exactMatches, partialMatches] = partition(
+        allMatches,
+        (item) => item.score === 1
+    )
+
+    const sortedPartialMatches = partialMatches.sort(
+        (a, b) => b.score - a.score
+    )
+
+    const combinedFilters = [
+        ...exactMatches.map((item) => item.filter),
+        ...(query ? [createQueryFilter(query)] : []),
+        ...sortedPartialMatches.map((item) => item.filter),
+    ]
+
+    return {
+        suggestions: combinedFilters,
+        unmatchedQuery,
+    }
+}
+
+export function createFilter(type: FilterType) {
+    return (name: string): Filter => ({ type, name })
 }
-export const analytics = new SiteAnalytics()
+
+export const createCountryFilter = createFilter(FilterType.COUNTRY)
+export const createTopicFilter = createFilter(FilterType.TOPIC)
+export const createQueryFilter = createFilter(FilterType.QUERY)
+
+export const SearchAutocompleteContext = createContext<
+    SearchAutocompleteContextType | undefined
+>(undefined)
+
+export function useSearchAutocomplete() {
+    const context = useContext(SearchAutocompleteContext)
+    if (context === undefined) {
+        throw new Error(
+            "useSearchAutocomplete must be used within a SearchAutocompleteContextProvider"
+        )
+    }
+    return context
+}
+
+/**
+ * Returns a click handler that focuses an input element when clicking directly
+ * on the target element (but not its children).
```

Comment:
I take it that this is so that if you click a remove topic/country button within `.search-bar` it doesn't focus trigger the :focus styles?

Reply:
It's to prevent clicks on the country selector to focus the input. I realized the case you describe should actually focus the input (clearing the form, removing a filter), so I updated accordingly (with a healthy dose of comments 🙂)